File: BOJ/[BOJ 1012] 유기농 배추.py
from collections import deque
def bfs(y, x):
    global  visited
    q = deque()
    q.append([y, x])
    visited[y][x] = 1
    while q:
        y, x = q.popleft()
        dy = [0,1,0,-1]
        dx = [1,0,-1,0]
        for i in range(4):
            ny, nx = y+dy[i], x+dx[i]
            if ny>n-1 or nx>m-1 or ny<0 or nx<0: continue
            if visited[ny][nx]==1: continue
            if arr[ny][nx]==0: continue
            visited[ny][nx]=1
            q.append([ny, nx])


t = int(input())
for _ in range(t):
    m, n, k = map(int, input().split())
    arr = [[0]*m for _ in range(n)]
    for i in range(k):
        x, y = map(int, input().split())
        arr[y][x]=1
    visited = [[0]*m for _ in range(n)]
    cnt = 0
    for i in range(n):
        for j in range(m):
            if arr[i][j]==1 and visited[i][j]==0:
                bfs(i, j)
                cnt += 1
    print(cnt)

File: BOJ/[BOJ 1012] 유기농 배추.js
const fs = require("fs");
const input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

const T = Number(input.splice(0, 1));

const BFS = (y, x, ground, visited, N, M) => {
  const q = [[y, x]];
  visited[y][x] = 1;
  const direct = [
    [0, 1],
    [0, -1],
    [1, 0],
    [-1, 0],
  ];
  while (q.length > 0) {
    const [y, x] = q.shift();
    for (let i = 0; i < 4; i++) {
      const ny = y + direct[i][0];
      const nx = x + direct[i][1];
      if (ny >= N || nx >= M || ny < 0 || nx < 0) continue;
      if (ground[ny][nx] === 0 || visited[ny][nx]) continue;
      visited[ny][nx] = 1;
      q.push([ny, nx]);
    }
  }
};

for (let i = 0; i < T; i++) {
  const [M, N, K] = input.splice(0, 1)[0].split(" ").map(Number);
  const lst = input.splice(0, K).map((v) => v.split(" ").map(Number));
  const ground = Array.from({ length: N }).map(() =>
    Array.from({ length: M }).fill(0)
  );
  const visited = JSON.parse(JSON.stringify(ground));

  for (const [x, y] of lst) ground[y][x] = 1;
  let count = 0;
  for (let i = 0; i < N; i++) {
    for (let j = 0; j < M; j++) {
      if (ground[i][j] === 1 && visited[i][j] === 0) {
        BFS(i, j, ground, visited, N, M);
        count += 1;
      }
    }
  }
  console.log(count);
}
